### ISSUE TITLE
Linux dynamic library linking failing when renv is activated (LD_LIBRARY_PATH / GDAL / terra)

### ISSUE THREAD

**Issue by dariomavec** [rstudio/renv#1796]
When installing the `terra` package on an RStudio Pro instance running on Red Hat, it will successfully install on a fresh session (linking to GDAL 2.4.4) but when installing in a renv 1.0.3 activated session it will fail, linking to GDAL 1.11.4 instead. I've replicated this behaviour on renv versions 1.0.2 and 1.0.0.

Is it possible that renv is affecting the `LD_LIBRARY_PATH`? Are there any additional options that should be set to help renv see linking libraries during compilation?

``` r
renv::install("terra")
#> # Downloading packages -------------------------------------------------------
#> - Downloading terra from CRAN ...               OK [file is up to date]
#> Successfully downloaded 1 package in 23 seconds.
#> 
#> The following package(s) will be installed:
#> - Rcpp  [1.0.12]
#> - terra [1.7-65]
#> These packages will be installed into "~/repos/temp/renv/library/R-4.2/x86_64-pc-linux-gnu".
#> 
#> # Installing packages --------------------------------------------------------
#> - Installing Rcpp ...                           OK [copied from cache]
#> - Installing terra ...                          FAILED
#> Error: Error installing package 'terra':
#> =================================
#> 
#> * installing *source* package ‘terra’ ...
#> ** package ‘terra’ successfully unpacked and MD5 sums checked
#> ** using staged installation
#> configure: CC: /opt/rh/devtoolset-12/root/usr/bin/gcc -std=gnu99
#> configure: CXX: /opt/rh/devtoolset-12/root/usr/bin/g++ -std=gnu++17
#> checking for gdal-config... /usr/bin/gdal-config
#> checking gdal-config usability... yes
#> configure: GDAL: 1.11.4
#> checking GDAL version >= 2.0.1... no
#> configure: error: terra is not compatible with GDAL versions below 2.0.1
#> ERROR: configuration failed for package ‘terra’
#> * removing ‘/home/mavecd@central.health/repos/temp/renv/library/R-4.2/x86_64-pc-linux-gnu/.renv/1/terra’
#> install of package 'terra' failed [error code 1]
```

``` r
Sys.getenv("LD_LIBRARY_PATH")
#> [1] "/opt/R/4.2.1/lib/R/lib:/usr/local/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.342.b07-1.el7_9.x86_64/jre/lib/amd64/server:/opt/R/4.2.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.342.b07-1.el7_9.x86_64/jre/lib/amd64/server:/usr/local/bin/gdal244/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-7.b13.el7.x86_64/jre/lib/amd64/server:/opt/teradata/client/15.10/lib64:/u01/drivers/SimbaGoogleBigQueryODBCConnector/SimbaODBCDriverforGoogleBigQuery64_3.0.2.1005/lib"
```

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: Red Hat
#> 
#> Matrix products: default
#> BLAS/LAPACK: /usr/lib64/libopenblasp-r0.3.3.so
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#>  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#>  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
#>  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#> [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> loaded via a namespace (and not attached):
#>  [1] digest_0.6.34     withr_3.0.0       R.methodsS3_1.8.2 lifecycle_1.0.4  
#>  [5] magrittr_2.0.3    reprex_2.1.0      evaluate_0.23     rlang_1.1.3      
#>  [9] cli_3.6.2         rstudioapi_0.15.0 fs_1.6.3          R.utils_2.12.3   
#> [13] R.oo_1.25.0       vctrs_0.6.5       styler_1.10.2     rmarkdown_2.25   
#> [17] tools_4.2.1       R.cache_0.16.0    glue_1.7.0        purrr_1.0.2      
#> [21] xfun_0.41         yaml_2.3.8        fastmap_1.1.1     compiler_4.2.1   
#> [25] htmltools_0.5.7   knitr_1.45
```

</details>

**Comment by kevinushey** [comment]
Can you please share the output of the following for your R session in each scenario?

```
Sys.getenv("LD_LIBRARY_PATH")
```

As well as:

```
system(paste0("cat /proc/", Sys.getpid(), "/environ | tr '\\0' '\\n' | grep LD_LIBRARY_PATH"))
```

For example, I see:

```
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/opt/R/4.2.3/lib/R/lib:/usr/local/lib:/usr/lib/jvm/java-11-openjdk-arm64/lib/server"
> system(paste0("cat /proc/", Sys.getpid(), "/environ | tr '\\0' '\\n' | grep LD_LIBRARY_PATH"))
LD_LIBRARY_PATH=/opt/R/4.2.3/lib/R/lib:/usr/local/lib:/usr/lib/jvm/java-11-openjdk-arm64/lib/server
```

**Comment by dariomavec** [comment]
Doesn't seem to be a difference between the two (provided below).

With `renv`:

```
> Sys.getenv("LD_LIBRARY_PATH")
[1] "/opt/R/4.2.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.342.b07-1.el7_9.x86_64/jre/lib/amd64/server:/usr/local/bin/gdal244/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-7.b13.el7.x86_64/jre/lib/amd64/server:/opt/teradata/client/15.10/lib64:/u01/drivers/SimbaGoogleBigQueryODBCConnector/SimbaODBCDriverforGoogleBigQuery64_3.0.2.1005/lib"
> system(paste0("cat /proc/", Sys.getpid(), "/environ | tr '\\0' '\\n' | grep LD_LIBRARY_PATH"))
LD_LIBRARY_PATH=/opt/R/4.2.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.342.b07-1.el7_9.x86_64/jre/lib/amd64/server:/usr/local/bin/gdal244/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-7.b13.el7.x86_64/jre/lib/amd64/server:/opt/teradata/client/15.10/lib64:/u01/drivers/SimbaGoogleBigQueryODBCConnector/SimbaODBCDriverforGoogleBigQuery64_3.0.2.1005/lib
```

Without `renv`:
```

> Sys.getenv("LD_LIBRARY_PATH")
[1] "/opt/R/4.2.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.342.b07-1.el7_9.x86_64/jre/lib/amd64/server:/usr/local/bin/gdal244/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-7.b13.el7.x86_64/jre/lib/amd64/server:/opt/teradata/client/15.10/lib64:/u01/drivers/SimbaGoogleBigQueryODBCConnector/SimbaODBCDriverforGoogleBigQuery64_3.0.2.1005/lib"
> system(paste0("cat /proc/", Sys.getpid(), "/environ | tr '\\0' '\\n' | grep LD_LIBRARY_PATH"))
LD_LIBRARY_PATH=/opt/R/4.2.1/lib/R/lib:/lib:/usr/local/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.342.b07-1.el7_9.x86_64/jre/lib/amd64/server:/usr/local/bin/gdal244/lib:/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.181-7.b13.el7.x86_64/jre/lib/amd64/server:/opt/teradata/client/15.10/lib64:/u01/drivers/SimbaGoogleBigQueryODBCConnector/SimbaODBCDriverforGoogleBigQuery64_3.0.2.1005/lib
```

**Comment by dariomavec** [comment]
After some more investigation, it appears this is being caused by the `PATH` envvar rather than the linking path.

Two versions of GDAL are installed on the server I am working from, one into `/usr/bin/` and one in a deeper directory that was included later on in the `PATH` variable.  Because `terra` uses `gdal-config` to check what version of GDAL is available, this is being triggered off the old executable and failing.

It seems that `renv` is appending `/usr/bin/` to the front of `PATH`.  Is there a way to prevent this behaviour, and what is the reason for it?

**PATH variables**
With `renv`:

```
> Sys.getenv("PATH")
[1] "/usr/bin:/u01/software/google-cloud-sdk/bin:/opt/teradata/client/16.20/bin:/usr/local/bin/gdal244/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/bin:/opt/openbugs/bin:/opt/openbugs/lib:/opt/puppetlabs/bin:/home/mavecd@central.health/.local/bin:/home/mavecd@central.health/bin:/usr/lib/rstudio-server/bin/quarto/bin:/usr/lib/rstudio-server/bin/postback"
```

Without `renv`:
```
> Sys.getenv("PATH")
[1] "/u01/software/google-cloud-sdk/bin:/opt/teradata/client/16.20/bin:/usr/local/bin/gdal244/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/bin:/opt/openbugs/bin:/opt/openbugs/lib:/opt/puppetlabs/bin:/home/mavecd@central.health/.local/bin:/home/mavecd@central.health/bin:/usr/lib/rstudio-server/bin/quarto/bin:/usr/lib/rstudio-server/bin/postback"
```

**Comment by kevinushey** [comment]
Could you try adding this code to your `.Rprofile` and let me know what you see?

```
trace(Sys.setenv, quote({
  if ("PATH" %in% names(list(...))) {
    print(rlang::trace_back())
  }
}))
```

My best guess is that perhaps this bit of code in `renv` is tripping:

https://github.com/rstudio/renv/blob/ae73cfcde5b9799f807634e14f780516f1193088/R/load.R#L482-L484

Have you configured `renv` to use a system installation of Python?

**Comment by kevinushey** [comment]
It would also be worth confirming whether you're using RStudio, as I believe RStudio had a similar issue in the past as well.

**Comment by dariomavec** [comment]
Trace below.

I am working on an RStudio Pro server instance. Is that what you mean by using RStudio?

```
Tracing function "Sys.setenv" in package "base"
[1] "Sys.setenv"
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = nsname) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(R_LIBS_SITE = .expand_R_libs_env_var(Sys.getenv("R_LIBS_SITE"))) on entry 
Tracing Sys.setenv(R_LIBS_USER = .expand_R_libs_env_var(Sys.getenv("R_LIBS_USER"))) on entry 
Tracing (new("functionWithTrace", .Data = function (...)  .... on entry 
Tracing (new("functionWithTrace", .Data = function (...)  .... on entry 
Tracing Sys.setenv(R_LIBS_SITE = .expand_R_libs_env_var(Sys.getenv("R_LIBS_SITE"))) on entry 
Tracing Sys.setenv(R_LIBS_USER = .expand_R_libs_env_var(Sys.getenv("R_LIBS_USER"))) on entry 
Tracing (new("functionWithTrace", .Data = function (...)  .... on entry 
Tracing (new("functionWithTrace", .Data = function (...)  .... on entry 
Tracing Sys.setenv(R_LIBS_USER = paste(libpaths, collapse = .Platform$path.sep)) on entry 
Tracing Sys.setenv(RETICULATE_MINICONDA_PYTHON_ENVPATH = envpath) on entry 
Tracing Sys.setenv(RENV_PROJECT = project) on entry 
- Project '~/xxx/xxx' loaded. [renv 1.0.3]
- The project is out-of-sync -- use `renv::status()` for details.
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = nsname) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = nsname) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = nsname) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = nsname) on entry 
Tracing Sys.setenv(`_R_NS_LOAD_` = package) on entry 
Tracing Sys.setenv(RSTUDIO_WINUTILS = "bin/winutils") on entry 
Tracing Sys.setenv(RSTUDIO_PANDOC = "/usr/lib/rstudio-server/bin/quarto/bin/tools",  .... on entry 
Tracing Sys.setenv(DISPLAY = ":0") on entry 
Tracing Sys.setenv(PATH = newPath) on entry 
    ▆
 1. └─.rs.python.initialize("/home/xxxxxx@xxx.xxx/")
 2.   └─.rs.prependToPath(pythonBin)
 3.     └─base::Sys.setenv(PATH = newPath)
Tracing Sys.setenv(RETICULATE_PYTHON_FALLBACK = pythonPath) on entry 
```

**Comment by kevinushey** [comment]
```
Tracing Sys.setenv(PATH = newPath) on entry 
    ▆
 1. └─.rs.python.initialize("/home/xxxxxx@xxx.xxx/")
 2.   └─.rs.prependToPath(pythonBin)
 3.     └─base::Sys.setenv(PATH = newPath)
```

This bit makes me think that RStudio could be the culprit. If you look within Tools -> Global Options... -> Python, you should see an option `[ ] Automatically activate project-local Python environments`. If that's selected, does unselecting it and restarting RStudio help?

<img width="642" alt="Screenshot 2024-01-30 at 11 29 30 PM" src="https://github.com/rstudio/renv/assets/1976582/dc6e0c13-13f4-4772-86d6-29acaca28e6b">

It may also help to check and configure in your Global Options + Project Options whether you've selected a system Python interpreter; changing that may also make a difference.

**Comment by dariomavec** [comment]
Amazing, thanks @kevinushey. This stopped "/usr/bin:" being appended to `PATH`.

Testing the various combinations of set-up, I found that the issue only occurs if both the Python interpreter isn't selected and the automatic activation of project-local Python environments is enabled (in screenshot below).  All other combos were fine. Essentially, make sure to have a Python environment activated.

![image](https://github.com/rstudio/renv/assets/4698363/d6041fd3-9506-461e-a428-b8d1f0aa55d9)

**Comment by kevinushey** [comment]
Great news -- thanks for verifying @dariomavec!

Are you able to confirm which version of RStudio you're using? I believe this should be resolved in the latest release (2023.12).

**Comment by dariomavec** [comment]
We are on the 2023.09 release (info below).  Good to hear this will be resolved for future users! :)

```
RStudio 2023.09.1+494.pro2 "Desert Sunflower" Release (80aa7dd23784dedd170aff0e7e1e1cdd1e99924b, 2023-10-16) for CentOS 7
Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36
```
